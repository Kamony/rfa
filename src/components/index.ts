export { CustomDialog, ConfirmationDialog } from "./dialogs";
export {
  Slider,
  Select,
  NumberInput,
  Switch,
  CheckBox,
  TextInput,
  Radio,
  SelectMulti,
} from "./form-components";
export {
  Url,
  Length,
  Positive,
  Negative,
  MoreThan,
  LessThan,
  Integer,
  Regex,
  Required,
  Email,
  Max,
  Min,
} from "./form-validators";
export { AttributesEditor } from "./attributes-editor";
export { DebugPanel } from "./debug-panel";
export { DraggableField } from "./draggable-field";
export { DraggableTag } from "./draggable-tag";
export { DropIndicator } from "./drop-indicator";
export { FormBuildElement } from "./form-build-item";
export { FormElementToolbox } from "./form-element-toolbox";
export { GroupChooserDialog } from "./group-chooser-dialog";
export { OptionsBuilder } from "./option-builder";
export { ValidationsEditor } from "./validations-editor";
