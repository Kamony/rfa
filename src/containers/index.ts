export { FormElementEditWindow } from './form-element-edit-window';
export { RfaSchemaTransferer } from './rfa-schema-transferer';
export { FormArchitect } from './form-architect';
export { DropArea } from './drop-area';
export { FieldBox } from './field-box';
export { FormElementsContainer } from './form-elements-container';
export { Previewer } from './previewer';
export { FormRenderer } from './form-renderer';

export { FormComponentWrapper } from './form-component-wrapper';
export { SortableWrapper } from './sortable-wrapper';
export { TabPanel } from './tab-panel';
