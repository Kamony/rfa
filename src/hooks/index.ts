export { useDataExporter } from "./useDataExporter";
export { useDialog } from "./useDialog";
export { useFieldCreator } from "./useFieldCreator";
export { useFormElementActions } from "./useFormElementActions";
export { useFormElements } from "./useFormElements";
export { useGrouping } from "./useGrouping";
export { useNameGenerator } from "./useNameGenerator";
export { useValidationCreator } from "./useValidationsCreator";
