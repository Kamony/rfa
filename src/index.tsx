export { FormRenderer } from './containers/form-renderer';
export { FormArchitect } from './containers/formArchitect';
export { RfaSchemaTransferer } from './containers/rfa-schema-transferer';

export type { IState } from './store/store';
export type { FormElement } from './model';
