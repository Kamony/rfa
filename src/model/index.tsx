import React from 'react';
import {
  CheckBoxOutlined as CheckBoxIcon,
  FormatColorTextOutlined as TextInputIcon,
  LinearScaleOutlined as SliderIcon,
  MenuOpenOutlined as SelectIcon,
  RadioButtonCheckedOutlined as RadioIcon,
  ToggleOnOutlined as SwitchIcon,
  Looks3Outlined as NumberIcon,
} from '@material-ui/icons';
import {
  CheckBox,
  NumberInput,
  Radio,
  Select,
  SelectMulti,
  Slider,
  Switch,
  TextInput,
} from '../components/form-components';
import {
  Email,
  Integer,
  Length,
  LessThan,
  Max,
  Min,
  MoreThan,
  Negative,
  Positive,
  Regex,
  Required,
  Url,
} from '../components/form-validators';

export type OptionType = {
  name: string;
};

export type AttributeValueType = string | number | boolean | string[];

export type AttributeSchema = {
  name: string;
  type: 'input' | 'select' | 'radio' | 'switch' | 'checkbox' | 'options'; // for editing purposes of attribute
  label: string;
  value?: AttributeValueType;
  options?: OptionType[];
};
export type ValidationType = 'string' | 'number' | 'array' | 'boolean'; //for setting validations

export type Validation = {
  type: string;
  params: any[];
};

export type FormElement = {
  name: string;
  label: string;
  description?: string;
  icon: React.ReactElement<any>;
  render: keyof typeof componentMapper;
  attributes: AttributeSchema[];
  validators?: (keyof typeof validatorsMapper)[];
  validationType: ValidationType;
  isReadOnly?: boolean;
  isCustomRegistered?: boolean;
};

export enum Components {
  INPUT = 'TextInput',
  NUMBER = 'NumberInput',
  CHECKBOX = 'CheckBox',
  RADIO = 'Radio',
  SELECT = 'Select',
  SELECTMULTI = 'SelectMulti',
  SWITCH = 'Switch',
  SLIDER = 'Slider',
}

export const componentMapper: Record<Components, React.FC<any>> = {
  [Components.INPUT]: TextInput,
  [Components.NUMBER]: NumberInput,
  [Components.CHECKBOX]: CheckBox,
  [Components.RADIO]: Radio,
  [Components.SELECT]: Select,
  [Components.SELECTMULTI]: SelectMulti,
  [Components.SWITCH]: Switch,
  [Components.SLIDER]: Slider,
};

export enum Validators {
  REQUIRED = 'required',
  MIN = 'min',
  MAX = 'max',
  LENGTH = 'length',
  REGEX = 'matches',
  EMAIL = 'email',
  URL = 'url',
  MORETHAN = 'moreThan',
  LESSTHAN = 'lessThan',
  POSITIVE = 'positive',
  NEGATIVE = 'negative',
  INTEGER = 'integer',
}

export const validatorsMapper: Record<Validators, React.FC<any>> = {
  [Validators.REQUIRED]: Required,
  [Validators.MIN]: Min,
  [Validators.MAX]: Max,
  [Validators.LENGTH]: Length,
  [Validators.REGEX]: Regex,
  [Validators.EMAIL]: Email,
  [Validators.URL]: Url,
  [Validators.MORETHAN]: MoreThan,
  [Validators.LESSTHAN]: LessThan,
  [Validators.POSITIVE]: Positive,
  [Validators.NEGATIVE]: Negative,
  [Validators.INTEGER]: Integer,
};

export const formElements: FormElement[] = [
  {
    name: 'text-input',
    label: 'Text',
    description: 'for names, addresses, passwords, urls...',
    icon: <TextInputIcon color={'secondary'} />,
    render: Components.INPUT,
    attributes: [
      { name: 'label', type: 'input', label: 'label', value: 'text label' },
      {
        name: 'placeholder',
        type: 'input',
        label: 'placeholder',
        value: 'text placeholder',
      },
      {
        name: 'helperText',
        type: 'input',
        label: 'helperText',
        value: 'helper text',
      },
      {
        name: 'type',
        type: 'select',
        label: 'type',
        options: [{ name: 'text' }, { name: 'password' }, { name: 'email' }],
        value: 'text',
      },
    ],
    validationType: 'string',
    validators: [
      Validators.REQUIRED,
      Validators.MIN,
      Validators.MAX,
      Validators.LENGTH,
      Validators.REGEX,
      Validators.EMAIL,
      Validators.URL,
    ],
  },
  {
    name: 'number-input',
    label: 'Number',
    description: 'for ages, sums, counts...',
    icon: <NumberIcon color={'secondary'} />,
    render: Components.NUMBER,
    attributes: [
      { name: 'label', type: 'input', label: 'label', value: 'number label' },
      {
        name: 'placeholder',
        type: 'input',
        label: 'placeholder',
        value: 'number placeholder',
      },
      {
        name: 'helperText',
        type: 'input',
        label: 'helperText',
        value: 'number helper text',
      },
    ],
    validationType: 'number',
    validators: [
      Validators.REQUIRED,
      Validators.MIN,
      Validators.MAX,
      Validators.LESSTHAN,
      Validators.MORETHAN,
      Validators.POSITIVE,
      Validators.NEGATIVE,
      Validators.INTEGER,
    ],
  },
  {
    name: 'checkbox',
    label: 'Checkbox',
    description: 'for selecting one or more options',
    icon: <CheckBoxIcon color={'secondary'} />,
    render: Components.CHECKBOX,
    attributes: [
      {
        name: 'label',
        type: 'input',
        label: 'label',
        value: 'checkboxes label',
      },
      {
        name: 'helperText',
        type: 'input',
        label: 'helperText',
        value: 'checkboxes helper text',
      },
      {
        type: 'options',
        name: 'options',
        label: 'manage options',
        options: [{ name: 'opt1' }, { name: 'opt2' }, { name: 'opt3' }],
      },
    ],
    validationType: 'array',
    validators: [Validators.REQUIRED],
  },
  {
    name: 'radio',
    label: 'Radio',
    description: 'for selecting one option',
    icon: <RadioIcon color={'secondary'} />,
    render: Components.RADIO,
    attributes: [
      { name: 'label', type: 'input', label: 'label', value: 'radio label' },
      {
        name: 'helperText',
        type: 'input',
        label: 'helperText',
        value: 'radio helper text',
      },
      {
        type: 'options',
        name: 'options',
        label: 'manage options',
        options: [{ name: 'opt1' }, { name: 'opt2' }, { name: 'opt3' }],
      },
    ],
    validationType: 'string',
    validators: [Validators.REQUIRED],
  },
  {
    name: 'select',
    label: 'Select',
    description: 'for selecting one from numerous options',
    icon: <SelectIcon color={'secondary'} />,
    render: Components.SELECT,
    attributes: [
      { name: 'label', type: 'input', label: 'label', value: 'select label' },
      {
        name: 'helperText',
        type: 'input',
        label: 'helperText',
        value: 'select helper text',
      },
      {
        type: 'options',
        name: 'options',
        label: 'manage options',
        options: [{ name: 'opt1' }, { name: 'opt2' }, { name: 'opt3' }],
      },
    ],
    validationType: 'string',
    validators: [Validators.REQUIRED],
  },
  {
    name: 'multiSelect',
    label: 'Select - multiple options',
    description: 'for selecting one or more from numerous options',
    icon: <SelectIcon color={'secondary'} />,
    render: Components.SELECTMULTI,
    attributes: [
      {
        name: 'label',
        type: 'input',
        label: 'label',
        value: 'multi select label',
      },
      {
        name: 'helperText',
        type: 'input',
        label: 'helperText',
        value: 'multi select helper text',
      },
      {
        type: 'options',
        name: 'options',
        label: 'manage options',
        options: [{ name: 'opt1' }, { name: 'opt2' }, { name: 'opt3' }],
      },
    ],
    validationType: 'array',
    validators: [Validators.REQUIRED],
  },
  {
    name: 'switch',
    label: 'Switch',
    description: 'for simple yes or no answers',
    icon: <SwitchIcon color={'secondary'} />,
    render: Components.SWITCH,
    attributes: [
      { name: 'label', type: 'input', label: 'label', value: 'switch label' },
      {
        name: 'helperText',
        type: 'input',
        label: 'helperText',
        value: 'switch helper text',
      },
    ],
    validationType: 'boolean',
  },
  {
    name: 'slider',
    label: 'Slider',
    description: 'for selecting a number interactively',
    icon: <SliderIcon color={'secondary'} />,
    render: Components.SLIDER,
    isCustomRegistered: true,
    attributes: [
      { name: 'label', type: 'input', label: 'label', value: 'slider label' },
      { name: 'min', type: 'input', label: 'Min Value', value: 0 },
      { name: 'max', type: 'input', label: 'Max Value', value: 100 },
      { name: 'step', type: 'input', label: 'Step', value: 1 },
      {
        name: 'helperText',
        type: 'input',
        label: 'helperText',
        value: 'slider helper text',
      },
    ],
    validationType: 'number',
    validators: [
      Validators.REQUIRED,
      Validators.MIN,
      Validators.MAX,
      Validators.LESSTHAN,
      Validators.MORETHAN,
      Validators.POSITIVE,
      Validators.NEGATIVE,
      Validators.INTEGER,
    ],
  },
];
