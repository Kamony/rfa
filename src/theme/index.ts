import { createMuiTheme } from '@material-ui/core';
//todo: add custom theme options
export const theme = createMuiTheme();
