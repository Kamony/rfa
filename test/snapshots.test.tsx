import React from 'react';
import renderer from 'react-test-renderer';
import {
  CheckBox,
  NumberInput,
  Radio,
  Switch,
  TextInput,
} from '../src/components/form-components';
import { Tree } from '../src';

test('Form Text', () => {
  const component = renderer.create(
    <TextInput register={jest.fn()} name={'text'} label={'text'} />
  );
  expect(component.toJSON()).toMatchSnapshot();
});

test('Form Number', () => {
  const component = renderer.create(
    <NumberInput register={jest.fn()} name={'text'} label={'text'} />
  );
  expect(component.toJSON()).toMatchSnapshot();
});

test('Form Checkbox', () => {
  const component = renderer.create(
    <CheckBox options={[]} register={jest.fn()} name={'text'} label={'text'} />
  );
  expect(component.toJSON()).toMatchSnapshot();
});

test('Form Radio', () => {
  const component = renderer.create(
    <Radio options={[]} register={jest.fn()} name={'text'} label={'text'} />
  );
  expect(component.toJSON()).toMatchSnapshot();
});

test('Form Switch', () => {
  const component = renderer.create(
    <Switch register={jest.fn()} name={'text'} label={'text'} />
  );
  expect(component.toJSON()).toMatchSnapshot();
});

test('Form Tree', () => {
  const component = renderer.create(<Tree node={{ name: 'test' }} />);
  expect(component.toJSON()).toMatchSnapshot();
});
