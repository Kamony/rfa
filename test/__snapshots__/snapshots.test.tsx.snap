// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Checkbox 1`] = `
<div
  className="MuiFormControl-root"
  style={
    Object {
      "width": "100%",
    }
  }
>
  <legend
    className="MuiFormLabel-root"
  >
    text
  </legend>
  <div
    className="MuiFormGroup-root"
  />
</div>
`;

exports[`Form Number 1`] = `
<div
  className="MuiFormControl-root"
  style={
    Object {
      "width": "100%",
    }
  }
>
  <div
    className="MuiFormControl-root MuiTextField-root"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink"
      data-shrink={true}
    >
      text
    </label>
    <div
      className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
      onClick={[Function]}
    >
      <input
        aria-invalid={false}
        autoFocus={false}
        className="MuiInputBase-input MuiInput-input"
        disabled={false}
        name="text"
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        required={false}
        type="number"
      />
    </div>
  </div>
</div>
`;

exports[`Form Radio 1`] = `
<div
  className="MuiFormControl-root"
  style={
    Object {
      "width": "100%",
    }
  }
>
  <legend
    className="MuiFormLabel-root"
  >
    text
  </legend>
  <div
    aria-label="text"
    className="MuiFormGroup-root"
    role="radiogroup"
  />
</div>
`;

exports[`Form Switch 1`] = `
<div
  className="MuiFormControl-root"
  style={
    Object {
      "width": "100%",
    }
  }
>
  <label
    className="MuiFormControlLabel-root"
  >
    <span
      className="MuiSwitch-root"
    >
      <span
        aria-disabled={false}
        className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-1 MuiSwitch-switchBase MuiSwitch-colorSecondary"
        onBlur={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={null}
      >
        <span
          className="MuiIconButton-label"
        >
          <input
            className="PrivateSwitchBase-input-4 MuiSwitch-input"
            disabled={false}
            name="text"
            onChange={[Function]}
            type="checkbox"
          />
          <span
            className="MuiSwitch-thumb"
          />
        </span>
      </span>
      <span
        className="MuiSwitch-track"
      />
    </span>
    <span
      className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
    >
      text
    </span>
  </label>
</div>
`;

exports[`Form Text 1`] = `
<div
  className="MuiFormControl-root"
  style={
    Object {
      "width": "100%",
    }
  }
>
  <div
    className="MuiFormControl-root MuiTextField-root"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink"
      data-shrink={true}
    >
      text
    </label>
    <div
      className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
      onClick={[Function]}
    >
      <input
        aria-invalid={false}
        autoFocus={false}
        className="MuiInputBase-input MuiInput-input"
        disabled={false}
        name="text"
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        required={false}
        type="text"
      />
    </div>
  </div>
</div>
`;

exports[`Form Tree 1`] = `
<div
  className="MuiContainer-root MuiContainer-maxWidthLg"
>
  <div
    style={
      Object {
        "border": "1px transparent solid",
      }
    }
  >
    <div>
      <div
        className="makeStyles-root-8"
        style={
          Object {
            "paddingLeft": 0,
          }
        }
      >
        <div
          className="makeStyles-root-12"
          onClick={[Function]}
        >
          <span
            className="MuiTypography-root makeStyles-nodeTitle-13 MuiTypography-caption MuiTypography-colorTextSecondary"
            style={
              Object {
                "visibility": "hidden",
              }
            }
          >
            test
          </span>
        </div>
        <div
          className="MuiPaper-root makeStyles-node-9 MuiPaper-outlined MuiPaper-rounded"
        >
          <div
            className="MuiBox-root MuiBox-root-14"
          >
            <h6
              className="MuiTypography-root MuiTypography-h6 MuiTypography-colorPrimary"
            >
              test
            </h6>
            <div />
          </div>
        </div>
        <div
          className="MuiCollapse-container MuiCollapse-hidden"
          style={
            Object {
              "minHeight": "0px",
            }
          }
        >
          <div
            className="MuiCollapse-wrapper"
          >
            <div
              className="MuiCollapse-wrapperInner"
            />
          </div>
        </div>
      </div>
      <hr
        className="MuiDivider-root"
      />
    </div>
  </div>
</div>
`;
